### PR TITLE
Fix `SearchIngredientScreen` Issues and Improve Generalization

### PR DESCRIPTION
**Pull request by AdriB02**
# Fix `SearchIngredientScreen` Issues and Improve Generalization

## Description
This pull request introduces enhancements, bug fixes, and refactors aimed at improving the functionality and user experience of ingredient-related screens within the recipe creation flow. The primary objective is to generalize the `SearchIngredientScreen` so it can be reused in other routes beyond recipe creation. Additionally, several bugs in the search functionality have been addressed, and the user interface has been enhanced for better usability.

#### Generalization and Feature Enhancements:

* Created a reusable `SearchIngredientScreen` for ingredient search across multiple routes.
* Introduced a new `ViewModel` interface to efficiently manage search behavior. This interface is designed with future scalability in mind and can be extended next week when the search functionality is implemented for the fridge. Currently, the `ViewModel` provides four methods to handle search operations and three `StateFlow` properties to manage the state of searched ingredients:
  - Methods
```kotlin
  /** Fetch an ingredient by scanning a barcode */
  fun fetchIngredient(barCode: Long)

  /** Clear the current ingredient after use */
  fun clearIngredient()

  /** Add an ingredient to the ingredient list */
  fun addIngredient(ingredient: Ingredient)

  /** Fetch ingredients by name */
  fun fetchIngredientByName(name: String)


  // The ingredient to search with the barcode
  val ingredient: StateFlow<Pair<Ingredient?, String?>>

  // The list of ingredients to display after fetching in the database
  val searchingIngredientList: StateFlow<List<Pair<Ingredient, String?>>>

  // The list of ingredient whe select
  val ingredientList: StateFlow<List<Pair<Ingredient, String?>>>
```

#### Bug Fixes:
* Added support for an optional quantity parameter when managing ingredients in recipes.
    * Updated ingredient to use Pair<Ingredient, String?> where the string represents quantity (nullable for cases without specified quantities) and use a Listt<Pair<Ingredient, String?>> to represents the fetched ingredient.
    * Ensured all ingredients have a quantity before allowing navigation to the next step, displaying an error message otherwise.

* Enforced a requirement for at least one ingredient in the recipe, with an error message shown for empty recipes.
* Fixed barcode scanner behavior to clear the scanned ingredient after use.
* Fixed bug where if multiple ingredient were added in the ingredient list, they are merged in only one ingredient and the quantities are added.
* Resolved a bug where duplicate ingredients in the list were incorrectly merged, summing their quantities.
* Addressed navigation issues:
   * Navigating back from search now reliably returns to the previous step in the recipe creation flow.
   * If the ingredient list is non-empty, the `IngredientListScreen` is displayed directly instead of the `RecipeIngredientScreen`.
   
#### UI enhancement:
* Added ingredient images to the ingredient list, sourced from the ingredient database.
* Improved visual consistency by reusing the same ingredient image as displayed in the barcode scanner.

#### Utils:
* Introduced a new utility function to display ingredient images within a styled box, utilized in both the barcode scanner and the ingredient list for consistent visuals:
```kotlin
fun IngredientImageBox(ingredient: Ingredient)
```
* Introduced a new `ConfirmationPopUp` to display a pop to ask for a user confirmation
```kotlin
fun ConfirmationPopUp(
    onConfirm: () -> Unit,
    onDismiss: () -> Unit,
    titleText: String,
    confirmationText: String,
    confirmationButtonText: String
)
``` 

#### Demo
* Scan barcode demo

https://github.com/user-attachments/assets/6e7d6138-fe27-4a3c-a3f5-229fd20a7c25

* Ingredient input fetching and listing demo

https://github.com/user-attachments/assets/4dd48224-1ec3-4e8c-906f-3a20a3c62665

### Why was this change made?
This change was made to enhance the reusability of the ingredient search feature across multiple routes, ensuring it can support diverse functionalities beyond recipe creation. Critical bugs affecting search functionality and user workflows have been addressed to provide a smoother and more reliable experience. Additionally, user experience has been improved with clearer error messages, better visuals, and more intuitive navigation. These updates also lay the groundwork for integrating similar features, such as fridge ingredient searches, in the future. The refactoring and modularity improvements make the codebase more maintainable and adaptable for future enhancements. Overall, this ensures a more robust and user-friendly system for managing ingredients.

## Checklist
- [x] Tests added.
- [x] Documentation updated.
- [x] All CI checks passed.
- [x] Linked issue/feature: #241

### CODE REVIEW

File: app/src/androidTest/java/com/android/sample/createRecipe/ListIngredientScreenTest.kt
Comment (CapraSnitzel) [inline app/src/androidTest/java/com/android/sample/createRecipe/ListIngredientScreenTest.kt:1]: Based on the code that you added, you could also add a test for when trying to go to the next step with an empty ingredients list. What do you think ?

File: app/src/androidTest/java/com/android/sample/createRecipe/IngredientSearchScreenTest.kt [hunk 1]
```diff
@@ -120,7 +119,6 @@ class IngredientSearchScreenTest {
 
     composeTestRule.onNodeWithTag(CONFIRMATION_BUTTON, useUnmergedTree = true).performClick()
     verify(mockNavigationActions).navigateTo(Screen.CREATE_RECIPE_LIST_INGREDIENTS)
-    assertEquals(ingredientViewModel.ingredientList.value, listOf(mockIngredients[0]))
```
Comment (CapraSnitzel) [inline app/src/androidTest/java/com/android/sample/createRecipe/IngredientSearchScreenTest.kt:123]: Is there a reason for why this verification was removed ?

File: app/src/test/java/com/android/sample/model/ingredient/IngredientViewModelTest.kt
Comment (CapraSnitzel) [inline app/src/test/java/com/android/sample/model/ingredient/IngredientViewModelTest.kt:1]: Based on the code added, you could test couple more things: the clear ingredients functionality, the case when we want to a null quantity with a non-null quantity and also when we want to add the quantities "g" and "500g". What do you think ?

**Review by CapraSnitzel — CHANGES_REQUESTED** [review]
Good job, you have made very important changes. You have the required minimum line coverage, which is very good. However, before approving your PR, there are some clarifications that need attention. Once that everything is clear, I will approve your PR.

File: app/src/main/java/com/android/sample/ui/createRecipe/IngredientSearchScreen.kt [hunk 2]
```diff
@@ -77,207 +21,15 @@ fun IngredientSearchScreen(
     navigationActions: NavigationActions,
     ingredientViewModel: IngredientViewModel,
 ) {
-  val listIngredient = ingredientViewModel.searchingIngredientList.collectAsState()
-  var showConfirmation by remember { mutableStateOf(DO_NOT_SHOW_CONFIRMATION) }
-  var selectedIngredient by remember { mutableStateOf<Ingredient?>(null) }
-  var isLoading by remember { mutableStateOf(INITIAL_LOADING_STATE) }
-
-  PlateSwipeScaffold(
+  SearchIngredientScreen(
```
[the remark this answers is not in the document]
Comment (AdriB02) [reply app/src/main/java/com/android/sample/ui/createRecipe/IngredientSearchScreen.kt:24]: Good suggestion! I'll make the changes.

File: app/src/main/java/com/android/sample/ui/searchIngredient/SearchIngredientScreen.kt [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+ * @param onDismiss the callback to invoke when the user dismisses the pop-up.
+ */
+@Composable
+private fun ConfirmationPopUp(
+    onConfirm: () -> Unit,
+    onDismiss: () -> Unit,
+    titleText: String,
+    confirmationText: String,
+    confirmationButtonText: String
+) {
```
[the remark this answers is not in the document]
Comment (AdriB02) [reply app/src/main/java/com/android/sample/ui/searchIngredient/SearchIngredientScreen.kt]: That's a great idea! I've already implemented it and added the `ConfirmationPopUp` to the `Utils` package in `ConfirmationPopUp.kt` for better reusability.

File: app/src/main/java/com/android/sample/ui/searchIngredient/SearchIngredientScreen.kt [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+                      for (ingredient in listIngredient.value) {
+                        IngredientItem(
+                            ingredient = ingredient.first,
+                            onClick = {
+                              selectedIngredient = ingredient.first
+                              showConfirmation = true
+                            })
+                      }
+                    }
+                  }
```
[the remark this answers is not in the document]
Comment (AdriB02) [reply app/src/main/java/com/android/sample/ui/searchIngredient/SearchIngredientScreen.kt]: Thanks! I've made some adjustments to the code you suggested, but I've implemented the main idea.

File: app/src/androidTest/java/com/android/sample/search/SearchIngredientScreenTest.kt [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+      onSuccess(emptyList())
+      null
+    }
+
+    composeTestRule
+        .onNodeWithTag(SEARCH_BAR, useUnmergedTree = true)
+        .performTextInput("UnknownIngredient")
+
+    composeTestRule.waitForIdle()
+  }
```
[the remark this answers is not in the document]
Comment (AdriB02) [reply app/src/androidTest/java/com/android/sample/search/SearchIngredientScreenTest.kt:207]: I've tested that no ingredients are displayed when the `TextInput` doesn't match any results. Additionally, I've added a new check to further improve the test.
```kotlin
    assertTrue(searchIngredientViewModel.searchingIngredientList.value.isEmpty())

```

File: app/src/main/java/com/android/sample/ui/createRecipe/ListIngredientScreen.kt [hunk 6]
```diff
@@ -134,130 +146,223 @@ fun IngredientListScreen(
...
+    navigationActions: NavigationActions
+) {
+  Box(
+      modifier = Modifier.fillMaxWidth().padding(PADDING_16.dp),
+      contentAlignment = Alignment.Center) {
+        Button(
+            onClick = {
+              if (ingredientList.isEmpty()) {
+                displayErrorIngredientMessage.value = true
+                return@Button
```
[the remark this answers is not in the document]
Comment (AdriB02) [reply app/src/main/java/com/android/sample/ui/createRecipe/ListIngredientScreen.kt:190]: In Kotlin, `return@Button` is a labeled return used inside a lambda. It makes the code exit from the `onClick` lambda when `ingredientList.isEmpty()` is true, without affecting the rest of the code outside the lambda.

It’s like saying "return from this button's click handler" and stop further execution of the lambda.

File: app/src/androidTest/java/com/android/sample/createRecipe/IngredientSearchScreenTest.kt (hunk 1 above)
Comment (AdriB02) [reply app/src/androidTest/java/com/android/sample/createRecipe/IngredientSearchScreenTest.kt:123]: Since the assertion no longer holds true after the modification, I’ve updated it with a new check.

File: app/src/test/java/com/android/sample/model/ingredient/IngredientViewModelTest.kt
Comment (AdriB02) [reply app/src/test/java/com/android/sample/model/ingredient/IngredientViewModelTest.kt:1]: I've added multiple tests to cover these cases. Let me know if you think of any other tests that should be included.

File: app/src/main/java/com/android/sample/ui/createRecipe/ListIngredientScreen.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (CapraSnitzel) [reply app/src/main/java/com/android/sample/ui/createRecipe/ListIngredientScreen.kt:190]: Thank you for the explenation. Now I understand better this part of the code.

File: app/src/test/java/com/android/sample/model/ingredient/IngredientViewModelTest.kt
Comment (CapraSnitzel) [reply app/src/test/java/com/android/sample/model/ingredient/IngredientViewModelTest.kt:1]: Looks good to me!

**Review by CapraSnitzel — APPROVED** [review]
